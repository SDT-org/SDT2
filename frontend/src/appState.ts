import React, { ErrorInfo } from "react";
import messages from "./messages";

export const performanceProfiles = {
  best: "Best",
  high: "High Performance",
  balanced: "Balanced",
  low: "Energy Saver",
} as const;
export const clusterMethods = ["Neighbor-Joining", "UPGMA", "None"] as const;

export type SaveableImageFormat = "png" | "jpeg" | "svg";

export type AppState = {
  view: "runner" | "loader" | "viewer";
  filename: string[];
  filetype: string;
  basename: string;
  progress: number;
  debug: boolean;
  sequences_count: number;
  alignment_output_path: string;
  export_path: string;
  performance_profiles: { [_: string]: number };
  stage: string;
  pair_progress: number;
  pair_count: number;
  estimated_time?: number;
  validation_error_id?: keyof typeof messages;
  client: {
    dataView: "heatmap" | "plot";
    cluster_method: (typeof clusterMethods)[number];
    performance_profile: keyof typeof performanceProfiles;
    error?: Error;
    errorInfo?: ErrorInfo;
    saveFormat: SaveableImageFormat;
    showExportModal: boolean;
  };
};

export const initialAppState: AppState = {
  view: "runner",
  filename: [""],
  filetype: "",
  basename: "",
  progress: 0,
  debug: false,
  sequences_count: 0,
  alignment_output_path: "",
  export_path: "",
  // These are just to make frontend easier to test, they get overwritten during the initial syncAppState
  performance_profiles: { best: 4, high: 3, balanced: 2, low: 1 },
  stage: "Preprocessing",
  pair_progress: 0,
  pair_count: 0,
  client: {
    dataView: "heatmap",
    cluster_method: "Neighbor-Joining",
    performance_profile: "best",
    saveFormat: "svg",
    showExportModal: false,
  },
};

export type SetAppState = React.Dispatch<React.SetStateAction<AppState>>;

export const AppStateContext = React.createContext<{
  appState: AppState;
  setAppState: SetAppState;
}>({
  appState: initialAppState,
  setAppState: () => undefined,
});

export const useAppState = () => {
  const context = React.useContext(AppStateContext);
  if (context === undefined) {
    throw new Error("useAppState must be used within an AppStateProvider");
  }
  return context;
};

export const syncAppState = (setAppState: SetAppState) => {
  if (!window.pywebview) {
    console.warn("Frontend-only mode detected, app state will not be synced.");
    return Promise.resolve();
  }

  return window.pywebview.api.get_state().then((data) => {
    setAppState((previous) => {
      return {
        ...previous,
        ...data,
      };
    });
  });
};

export default useAppState;
