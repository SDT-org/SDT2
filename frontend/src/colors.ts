export type ColorString =
  | `hsl(${number}, ${number}%, ${number}%)`
  | `hsla(${number}, ${number}%, ${number}%, ${number})`;

type ColorName =
  | "White"
  | "Black"
  | "Tomato"
  | "LightBlue"
  | "LightGreen"
  | "Purple"
  | "Orange"
  | "Yellow"
  | "Cyan"
  | "Teal"
  | "Magenta"
  | "Brown"
  | "Lime"
  | "Coral"
  | "Turquoise"
  | "Indigo"
  | "Violet"
  | "Lavender"
  | "Peach"
  | "SkyBlue"
  | "Olive"
  | "Tan"
  | "Salmon"
  | "Maroon"
  | "Navy"
  | "Khaki"
  | "Periwinkle"
  | "Mint"
  | "Azure"
  | "Chartreuse"
  | "Goldrod"
  | "SlateBlue"
  | "LightSeaGreen"
  | "DarkCyan"
  | "RosyBrown"
  | "PaleVioletRed"
  | "DeepPink"
  | "DarkOrange"
  | "Crimson"
  | "LightSalmon"
  | "Orchid"
  | "Thistle"
  | "DarkKhaki"
  | "LightCoral"
  | "MediumOrchid"
  | "None";

export const Colors: Record<ColorName, ColorString> = {
  White: "hsl(0, 0%, 100%)",
  Black: "hsl(0, 0%, 0%)",
  Tomato: "hsl(9, 100%, 64%)",
  LightBlue: "hsl(195, 53%, 79%)",
  LightGreen: "hsl(120, 73%, 75%)",
  Purple: "hsl(300, 47%, 75%)",
  Orange: "hsl(39, 100%, 50%)",
  Yellow: "hsl(51, 100%, 50%)",
  Cyan: "hsl(180, 100%, 50%)",
  Teal: "hsl(180, 100%, 25%)",
  Magenta: "hsl(300, 100%, 50%)",
  Brown: "hsl(25, 76%, 31%)",
  Lime: "hsl(120, 100%, 50%)",
  Coral: "hsl(16, 100%, 66%)",
  Turquoise: "hsl(174, 72%, 56%)",
  Indigo: "hsl(275, 100%, 25%)",
  Violet: "hsl(300, 76%, 72%)",
  Lavender: "hsl(240, 67%, 94%)",
  Peach: "hsl(28, 100%, 86%)",
  SkyBlue: "hsl(197, 71%, 73%)",
  Olive: "hsl(60, 100%, 25%)",
  Tan: "hsl(34, 44%, 69%)",
  Salmon: "hsl(6, 93%, 71%)",
  Maroon: "hsl(0, 100%, 25%)",
  Navy: "hsl(240, 100%, 25%)",
  Khaki: "hsl(54, 77%, 75%)",
  Periwinkle: "hsl(54, 77%, 75%)",
  Mint: "hsl(150, 100%, 98%)",
  Azure: "hsl(180, 100%, 97%)",
  Chartreuse: "hsl(90, 100%, 50%)",
  Goldrod: "hsl(43, 74%, 49%)",
  SlateBlue: "hsl(248, 53%, 58%)",
  LightSeaGreen: "hsl(177, 70%, 41%)",
  DarkCyan: "hsl(180, 100%, 27%)",
  RosyBrown: "hsl(0, 25%, 65%)",
  PaleVioletRed: "hsl(340, 60%, 65%)",
  DeepPink: "hsl(328, 100%, 54%)",
  DarkOrange: "hsl(33, 100%, 50%)",
  Crimson: "hsl(348, 83%, 47%)",
  LightSalmon: "hsl(17, 100%, 74%)",
  Orchid: "hsl(302, 59%, 65%)",
  Thistle: "hsl(300, 24%, 80%)",
  DarkKhaki: "hsl(56, 38%, 58%)",
  LightCoral: "hsl(0, 79%, 72%)",
  MediumOrchid: "hsl(288, 59%, 58%)",
  None: "hsla(0, 0%, 0%, 0)",
} as const;
