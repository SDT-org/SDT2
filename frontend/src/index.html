<!DOCTYPE html>
<html style="background: #efefef">
  <head>
    <title>SDT2</title>
    <meta name="format-detection" content="telephone=no"/>
    <link href="index.scss" type="text/css">
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="index.tsx"></script>
  </body>
</html>
