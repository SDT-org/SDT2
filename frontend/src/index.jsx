import React from 'react';
import { createRoot } from 'react-dom/client';
import './index.scss';
import { App } from './components/App';

const container = document.getElementById('app');
const root = createRoot(container);
root.render(<App />);
