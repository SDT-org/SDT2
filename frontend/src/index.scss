/* https://www.joshwcomeau.com/css/custom-css-reset/ */

/*
  1. Use a more-intuitive box-sizing model.
*/
*, *::before, *::after {
  box-sizing: border-box;
}
/*
  2. Remove default margin
*/
* {
  margin: 0;
}
html, body, #app {
  height: 100%;
}
/*
  Typographic tweaks!
  3. Add accessible line-height
  4. Improve text rendering
*/
body {
  line-height: 1.5;
  -webkit-font-smoothing: antialiased;
}
/*
  5. Improve media defaults
*/
img, picture, video, canvas, svg {
  display: inherit;
  max-width: 100%;
}
/*
  6. Remove built-in form typography styles
*/
input, button, textarea, select {
  font: inherit;
}
/*
  7. Avoid text overflows
*/
p, h1, h2, h3, h4, h5, h6 {
  overflow-wrap: break-word;
}
/*
  8. Create a root stacking context
*/
#app, #root, #__next {
  isolation: isolate;
}


/* BASE STYLES */

html { font-family: 'Inter', sans-serif; }
@supports (font-variation-settings: normal) {
  html { font-family: 'Inter var', sans-serif; }
}

@import './styles/fonts.scss';
@import './styles/app.scss';
@import './styles/modal.scss';
