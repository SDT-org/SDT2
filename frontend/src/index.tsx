import { createRoot } from "react-dom/client";
import "./index.scss";
import { StrictMode } from "react";
import { App } from "./components/App";

const container = document.getElementById("app");
if (!container) {
  throw new Error("could not find container to mount app");
}

const root = createRoot(container);
root.render(
  <StrictMode>
    <App />
  </StrictMode>,
);
