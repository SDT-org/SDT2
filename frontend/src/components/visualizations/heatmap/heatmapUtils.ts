import tinycolor from "tinycolor2";
import type { ColorScaleArray } from "../../../colorScales";
import { createD3ColorScale, distinctColor } from "../../../colors";
import type {
  ClusterDataItem,
  HeatmapData,
  HeatmapSettings,
} from "../../../plotTypes";

export const getPlotMetrics = (width: number, height: number) => {
  // WIP
  return {
    width: width,
    height: height,
  };
};

export const getCellMetrics = (
  cellSize: number,
  cellSpace: number,
  characterCount: number,
) => {
  if (characterCount <= 0) throw new Error("characterCount must be > 0");

  const CHARACTER_WIDTH = 0.6; // Approximate width for Roboto Mono @ 10px
  const USABLE_SPACE_RATIO = 0.8;
  const MIN_FONT_SIZE = 0.25;
  const MAX_FONT_SIZE = 20;

  const scaledCellSpace = cellSpace * (cellSize / (cellSize + 20));
  const spacedCellSize = Math.max(1, cellSize - scaledCellSpace);
  const scaledFontSize = Math.min(
    MAX_FONT_SIZE,
    Math.max(
      MIN_FONT_SIZE,
      (spacedCellSize * USABLE_SPACE_RATIO) /
        (characterCount * CHARACTER_WIDTH),
    ),
  );

  return {
    cellSize: spacedCellSize,
    cellSpace: scaledCellSpace,
    cellOffset: scaledCellSpace / 2,
    fontSize: scaledFontSize,
    textOffset: spacedCellSize / 2,
  };
};

export const formatHeatmapData = (
  data: HeatmapData,
  settings: Pick<
    HeatmapSettings,
    | "colorScaleKey"
    | "vmax"
    | "vmin"
    | "annotation_rounding"
    | "hideValuesBelow"
    | "hideValuesBelowEnabled"
  >,
  colorScale: ColorScaleArray,
) => {
  const colorFn = createD3ColorScale(
    colorScale,
    settings.colorScaleKey === "Discrete",
    settings.vmax,
    settings.vmin,
  );

  const emptyColor = "#f5f5f5";
  const blackColor = "#000";
  const whiteColor = "#fff";
  const hideValuesBelowEnabled = settings.hideValuesBelowEnabled;
  const hideValuesBelow = settings.hideValuesBelow;

  const colorCache = new Map<string, { bg: string; fg: string }>();

  // Pre-allocate result array to avoid multiple array resizing
  let totalCells = 0;
  for (let y = 0; y < data.length; y++) {
    const row = data[y];
    if (!row) continue;
    for (let x = 0; x < row.length; x++) {
      if (row[x] === 0 || Number(row[x])) {
        totalCells++;
      }
    }
  }

  const result = new Array(totalCells);
  let resultIndex = 0;

  for (let y = 0; y < data.length; y++) {
    const row = data[y];
    if (!row) continue;
    let actualX = 0;

    for (let x = 0; x < row.length; x++) {
      const datum = row[x];

      // can skip empty/null values early
      if (!(datum === 0 || Number(datum))) {
        continue;
      }

      const value = datum as number;

      const shouldHide = hideValuesBelowEnabled && value < hideValuesBelow;
      const effectiveValue = shouldHide ? 0 : value;

      let displayValue: string;
      if (effectiveValue === 100) {
        displayValue = "100";
      } else if (effectiveValue === 0) {
        displayValue = "";
      } else {
        displayValue = effectiveValue.toFixed(settings.annotation_rounding);
      }

      let backgroundColor: string;
      let foregroundColor: string;

      if (displayValue === "") {
        backgroundColor = emptyColor;
        foregroundColor = blackColor;
      } else {
        const colorKey = effectiveValue.toString();
        const colors = colorCache.get(colorKey);

        if (!colors) {
          backgroundColor = colorFn(effectiveValue);
          foregroundColor = tinycolor(backgroundColor).isLight()
            ? blackColor
            : whiteColor;
          colorCache.set(colorKey, {
            bg: backgroundColor,
            fg: foregroundColor,
          });
        } else {
          backgroundColor = colors.bg;
          foregroundColor = colors.fg;
        }
      }

      result[resultIndex] = {
        x: actualX,
        y,
        value,
        displayValue,
        backgroundColor,
        foregroundColor,
      };

      resultIndex++;
      actualX++;
    }
  }

  return result;
};

export const formatClustermapData = (
  data: HeatmapData,
  tickText: string[],
  clusterData?: ClusterDataItem[],
) => {
  const clusterMap = new Map();
  const originalClusterMap = new Map();
  if (clusterData) {
    for (const item of clusterData) {
      clusterMap.set(item.id, item.cluster);
      // Use original_cluster if available, otherwise use cluster
      originalClusterMap.set(item.id, item.original_cluster ?? item.cluster);
    }
  }

  const colorCache = new Map();

  const result = data.flatMap((row, y) =>
    row
      .filter((datum) => datum === 0 || Number(datum))
      .map((value, x) => {
        const clusterX = clusterMap.get(tickText[x]);
        const clusterY = clusterMap.get(tickText[y]);
        const originalClusterX = originalClusterMap.get(tickText[x]);
        const originalClusterY = originalClusterMap.get(tickText[y]);

        const clusterMatch =
          clusterX !== undefined &&
          clusterY !== undefined &&
          clusterX === clusterY;

        const clusterGroup = clusterMatch ? clusterX : null;
        const originalClusterGroup =
          clusterMatch && originalClusterX === originalClusterY
            ? originalClusterX
            : null;

        let backgroundColor = "rgb(245, 245, 245)";
        if (clusterGroup && originalClusterGroup) {
          if (!colorCache.has(clusterGroup)) {
            // Use original cluster number for color generation
            colorCache.set(clusterGroup, distinctColor(originalClusterGroup));
          }
          backgroundColor = colorCache.get(clusterGroup);
        }

        let foregroundColor = "rgb(0, 0, 0)";
        if (!colorCache.has(`${backgroundColor}-fg`)) {
          foregroundColor = tinycolor(backgroundColor).isLight()
            ? "#000"
            : "#fff";
          colorCache.set(`${backgroundColor}-fg`, foregroundColor);
        } else {
          foregroundColor = colorCache.get(`${backgroundColor}-fg`);
        }

        const roundedValue = (value as number).toFixed(2);
        const displayValue =
          value === 100 ? "100" : value === 0 ? "" : roundedValue.toString();

        return {
          x,
          y,
          value: value as number,
          displayValue,
          backgroundColor,
          foregroundColor,
        };
      }),
  );

  return result;
};
