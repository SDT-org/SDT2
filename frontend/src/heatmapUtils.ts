import tinycolor from "tinycolor2";
import type { ColorScaleArray } from "./colorScales";
import { createD3ColorScale, distinctColor } from "./colors";
import type {
  ClusterDataItem,
  HeatmapData,
  HeatmapSettings,
} from "./plotTypes";

export const getPlotMetrics = (width: number, height: number) => {
  // WIP
  return {
    width: width,
    height: height,
  };
};

export const getCellMetrics = (
  cellSize: number,
  cellSpace: number,
  characterCount: number,
) => {
  if (characterCount <= 0) throw new Error("characterCount must be > 0");

  const CHARACTER_WIDTH = 0.6; // Approximate width for Roboto Mono @ 10px
  const USABLE_SPACE_RATIO = 0.8;
  const MIN_FONT_SIZE = 0.25;
  const MAX_FONT_SIZE = 20;

  const scaledCellSpace = cellSpace * (cellSize / (cellSize + 20));
  const spacedCellSize = Math.max(1, cellSize - scaledCellSpace);
  const scaledFontSize = Math.min(
    MAX_FONT_SIZE,
    Math.max(
      MIN_FONT_SIZE,
      (spacedCellSize * USABLE_SPACE_RATIO) /
        (characterCount * CHARACTER_WIDTH),
    ),
  );

  return {
    cellSize: spacedCellSize,
    cellSpace: scaledCellSpace,
    cellOffset: scaledCellSpace / 2,
    fontSize: scaledFontSize,
    textOffset: spacedCellSize / 2,
  };
};

export const formatHeatmapData = (
  data: HeatmapData,
  settings: Pick<
    HeatmapSettings,
    "colorScaleKey" | "vmax" | "vmin" | "annotation_rounding"
  >,
  colorScale: ColorScaleArray,
  metaData?: { run?: { analysis_method?: string } },
) => {
  const colorFn = createD3ColorScale(
    colorScale,
    settings.colorScaleKey === "Discrete",
    settings.vmax,
    settings.vmin,
  );

  const isLzani = metaData?.run?.analysis_method === "lzani";
  const lzaniThreshold = 70;

  return data.flatMap((row, y) =>
    row
      .filter((datum) => datum === 0 || Number(datum))
      .map((value, x) => {
        // For LZ-ANI, treat values below 70% as unaligned (0%)
        const effectiveValue =
          isLzani && (value as number) < lzaniThreshold ? 0 : value;

        const roundedValue = (effectiveValue as number).toFixed(
          settings.annotation_rounding,
        );
        const displayValue =
          effectiveValue === 100
            ? "100"
            : effectiveValue === 0
              ? ""
              : roundedValue.toString();

        const backgroundColor =
          displayValue === "" ? "#f5f5f5" : colorFn(Number(effectiveValue));
        const foregroundColor = tinycolor(backgroundColor).isLight()
          ? "#000"
          : "#fff";

        return {
          x,
          y,
          value: value as number,
          displayValue,
          backgroundColor,
          foregroundColor,
        };
      }),
  );
};

export const formatClustermapData = (
  data: HeatmapData,
  tickText: string[],
  clusterData?: ClusterDataItem[],
) => {
  const clusterMap = new Map();
  const originalClusterMap = new Map();
  if (clusterData) {
    for (const item of clusterData) {
      clusterMap.set(item.id, item.cluster);
      // Use original_cluster if available, otherwise use cluster
      originalClusterMap.set(item.id, item.original_cluster ?? item.cluster);
    }
  }

  const colorCache = new Map();

  const result = data.flatMap((row, y) =>
    row
      .filter((datum) => datum === 0 || Number(datum))
      .map((value, x) => {
        const clusterX = clusterMap.get(tickText[x]);
        const clusterY = clusterMap.get(tickText[y]);
        const originalClusterX = originalClusterMap.get(tickText[x]);
        const originalClusterY = originalClusterMap.get(tickText[y]);

        const clusterMatch =
          clusterX !== undefined &&
          clusterY !== undefined &&
          clusterX === clusterY;

        const clusterGroup = clusterMatch ? clusterX : null;
        const originalClusterGroup =
          clusterMatch && originalClusterX === originalClusterY
            ? originalClusterX
            : null;

        let backgroundColor = "rgb(245, 245, 245)";
        if (clusterGroup && originalClusterGroup) {
          if (!colorCache.has(clusterGroup)) {
            // Use original cluster number for color generation
            colorCache.set(clusterGroup, distinctColor(originalClusterGroup));
          }
          backgroundColor = colorCache.get(clusterGroup);
        }

        let foregroundColor = "rgb(0, 0, 0)";
        if (!colorCache.has(`${backgroundColor}-fg`)) {
          foregroundColor = tinycolor(backgroundColor).isLight()
            ? "#000"
            : "#fff";
          colorCache.set(`${backgroundColor}-fg`, foregroundColor);
        } else {
          foregroundColor = colorCache.get(`${backgroundColor}-fg`);
        }

        const roundedValue = (value as number).toFixed(2);
        const displayValue =
          value === 100 ? "100" : value === 0 ? "" : roundedValue.toString();

        return {
          x,
          y,
          value: value as number,
          displayValue,
          backgroundColor,
          foregroundColor,
        };
      }),
  );

  return result;
};
