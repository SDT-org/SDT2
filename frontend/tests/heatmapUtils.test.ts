import { describe, expect, it } from "bun:test";
import { getCellMetrics } from "../src/heatmapUtils";

describe("getCellMetrics", () => {
  it("should return the correct metrics", () => {
    expect(getCellMetrics(20, 1, 5)).toEqual({
      cellSize: 19.5,
      cellSpace: 0.5,
      cellOffset: 0.25,
      fontSize: 5.2,
      textOffset: 9.75,
    });
  });
  it("should throw an error if characterCount is less than or equal to 0", () => {
    expect(() => getCellMetrics(10, 2, 0)).toThrowError(
      "characterCount must be > 0",
    );
    expect(() => getCellMetrics(10, 2, -1)).toThrowError(
      "characterCount must be > 0",
    );
  });
});
