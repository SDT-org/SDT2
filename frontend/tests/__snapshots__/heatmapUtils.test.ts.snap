// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`formatHeatmapData should return formatted data 1`] = `
[
  {
    "backgroundColor": "rgb(255, 255, 255)",
    "foregroundColor": "#000",
    "value": 100,
    "x": 0,
    "y": 0,
  },
  {
    "backgroundColor": "rgb(179, 179, 179)",
    "foregroundColor": "#000",
    "value": 73,
    "x": 0,
    "y": 1,
  },
  {
    "backgroundColor": "rgb(255, 255, 255)",
    "foregroundColor": "#000",
    "value": 100,
    "x": 1,
    "y": 1,
  },
  {
    "backgroundColor": "rgb(179, 179, 179)",
    "foregroundColor": "#000",
    "value": 73,
    "x": 0,
    "y": 2,
  },
  {
    "backgroundColor": "rgb(248, 248, 248)",
    "foregroundColor": "#000",
    "value": 98,
    "x": 1,
    "y": 2,
  },
  {
    "backgroundColor": "rgb(255, 255, 255)",
    "foregroundColor": "#000",
    "value": 100,
    "x": 2,
    "y": 2,
  },
  {
    "backgroundColor": "rgb(0, 0, 0)",
    "foregroundColor": "#fff",
    "value": 8,
    "x": 0,
    "y": 3,
  },
  {
    "backgroundColor": "rgb(163, 163, 163)",
    "foregroundColor": "#000",
    "value": 67,
    "x": 1,
    "y": 3,
  },
  {
    "backgroundColor": "rgb(163, 163, 163)",
    "foregroundColor": "#000",
    "value": 67,
    "x": 2,
    "y": 3,
  },
  {
    "backgroundColor": "rgb(255, 255, 255)",
    "foregroundColor": "#000",
    "value": 100,
    "x": 3,
    "y": 3,
  },
]
`;

exports[`formatClustermapData should return formatted data 1`] = `
[
  {
    "backgroundColor": "hsl(137.5, 85%, 50%)",
    "foregroundColor": "#000",
    "value": 100,
    "x": 0,
    "y": 0,
  },
  {
    "backgroundColor": "rgb(245, 245, 245)",
    "foregroundColor": "#000",
    "value": 72.6,
    "x": 0,
    "y": 1,
  },
  {
    "backgroundColor": "hsl(275, 85%, 50%)",
    "foregroundColor": "#fff",
    "value": 100,
    "x": 1,
    "y": 1,
  },
  {
    "backgroundColor": "rgb(245, 245, 245)",
    "foregroundColor": "#000",
    "value": 72.62,
    "x": 0,
    "y": 2,
  },
  {
    "backgroundColor": "hsl(275, 85%, 50%)",
    "foregroundColor": "#fff",
    "value": 97.61,
    "x": 1,
    "y": 2,
  },
  {
    "backgroundColor": "hsl(275, 85%, 50%)",
    "foregroundColor": "#fff",
    "value": 100,
    "x": 2,
    "y": 2,
  },
  {
    "backgroundColor": "rgb(245, 245, 245)",
    "foregroundColor": "#000",
    "value": 8.2,
    "x": 0,
    "y": 3,
  },
  {
    "backgroundColor": "rgb(245, 245, 245)",
    "foregroundColor": "#000",
    "value": 66.99,
    "x": 1,
    "y": 3,
  },
  {
    "backgroundColor": "rgb(245, 245, 245)",
    "foregroundColor": "#000",
    "value": 66.88,
    "x": 2,
    "y": 3,
  },
  {
    "backgroundColor": "hsl(52.5, 85%, 50%)",
    "foregroundColor": "#000",
    "value": 100,
    "x": 3,
    "y": 3,
  },
]
`;
